### PR TITLE
providers: add support for Azure

### PR DESCRIPTION
**Pull request by crawford**

### CODE REVIEW

File: build [hunk 1]
```diff
@@ -4,7 +4,7 @@ NAME="coreos-metadata"
 ORG_PATH="github.com/coreos"
 REPO_PATH="${ORG_PATH}/${NAME}"
 VERSION=$(git describe --dirty)
-GLDFLAGS="-X main.version \"${VERSION}\""
+GLDFLAGS="-X main.version=\"${VERSION}\""
```
Comment (jonboulle) [inline build]: Huh. We have this nonsense: https://github.com/coreos/rkt/blob/master/configure.ac#L27

Comment (vcaputo) [inline build]: doesn't = work everywhere?

Comment (jonboulle) [inline build]: I assumed not based on
https://github.com/coreos/rkt/commit/2b1a60db047f5b2a408586e01678c461446a5c80

Comment (crawford) [inline build]: It built in Travis...

File: src/providers/azure/azure.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+	if err := assertFabricCompatible(addr, FabricProtocolVersion); err != nil {
+		return config.Metadata{}, err
+	}
+
+	return fetchSharedConfig(addr)
+}
+
+func getFabricAddress() (net.IP, error) {
+	lease, err := os.Open("/run/systemd/netif/leases/2")
```
Comment (vcaputo) [inline src/providers/azure/azure.go]: this looks like it may be assuming too much

File: build (hunk 1 above)
Comment (jonboulle) [inline build]: I dunno, /cc @krnowak

Comment (crawford) [inline build]: Wow, you've got to be fucking kidding me. It silently fails. Go strikes again!

File: src/providers/azure/azure.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	client := retry.Client{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 5,
+		MaxAttempts:    10,
+		Header: map[string][]string{
+			"x-ms-agent-name": {AgentName},
+			"x-ms-version":    {FabricProtocolVersion},
+			"Content-Type":    {"text/xml; charset=utf-8"},
+		},
+	}
```
Comment (vcaputo) [inline src/providers/azure/azure.go:130]: Why are you repeating yourself?  Stick this in a function which returns the populated retry.Client and reuse with above, they're doing identical things.

File: src/providers/azure/azure.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+				Id             string `xml:"id,attr"`
+				Address        string `xml:"address,attr"`
+				InputEndpoints struct {
+					Endpoints []struct {
+						LoadBalancedPublicAddress string `xml:"loadBalancedPublicAddress,attr"`
+					} `xml:"Endpoint"`
+				}
+			} `xml:"Instance"`
+		}
+	}{}
```
Comment (vcaputo) [inline src/providers/azure/azure.go:173]: It just feels icky to have these guys which amount to schema sprinkled inline with the code, but I don't know what's "idiomatic".

File: src/providers/azure/azure.go (hunk 3 above)
Comment (crawford) [inline src/providers/azure/azure.go:130]: Unrepeated.

**Comment by vcaputo** [comment]
lgtm overall

File: src/providers/azure/azure.go (hunk 2 above)
Comment (crawford) [inline src/providers/azure/azure.go]: What about `net.InterfaceByName("eth0").Index()`?

Comment (crawford) [inline src/providers/azure/azure.go]: I could also just test all leases.

File: src/providers/azure/azure.go (hunk 4 above)
Comment (crawford) [inline src/providers/azure/azure.go:173]: I don't know either, but I figured I do it since I only ever use it once.

**Comment by vcaputo** [comment]
lgtm now and forever

**Pull request merged**